### DISCUSSION TITLE
Proposal: client-side upsert

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#753] [Q&A]
Referring to https://www.sqlite.org/lang_UPSERT.html, how about supporting the simple cases like this:

```sql
CREATE TABLE phonebook(id INT PRIMARY KEY, name TEXT, number TEXT);
CREATE UNIQUE INDEX idx_phone_number
ON phonebook (number);

INSERT INTO phonebook(name,number) VALUES('Alice','704-555-1212')
  ON CONFLICT(number) DO UPDATE SET name='Alice'
```

```ts
const name = `Alice`
const number = `704-555-1212`
client.records.upsert(`phonebook`, { name, number }, { number })
```

Supporting `upsert` would avoid race conditions for simple operations because every SQLite statement happens inside its own transaction.

Thoughts?

**Comment by ganigeorgiev** [comment]
_**Edit**: clientside upsert support will be available with the next v0.23.0 release (see https://github.com/pocketbase/pocketbase/discussions/753#discussioncomment-10420268)_

---

Unfortunately, adding support for upsert in PocketBase will be a little more complicated because we'll need to check also the create or update API rule depending on what operation is executed, aka. I don't think we can use directly the `UPSERT` SQLite statement. Additionally, without a primary key (id) it is a little ambiguous which record will be updated (the collection may not have unique constraint or the unique constraint may be partial).

Race conditions related to a single select + create/update/delete are rarely an issue since we are not using autoincrement/counter fields and wrapping those in a transaction actually may result in poorer performance and eventually to SQLITE_BUSY error.

For now this is not planned, but bulk transactional action for create/update/delete operations is in my todo and it will be implemented in the near future.

**Reply by lukepighetti** [reply]
my take:

require an id to be provided for upsert

if that record (by id) does not exist, create, use create rule
if that record (by id) exists, update, use update rule

**Reply by ganigeorgiev** [reply]
The problem here is not that we can't decide which API rule to apply but that we cannot use the SQLite `UPSERT` clause because we can't be sure whether the record is new or existing before executing the query. In addition to the API rules, we also need to know upfront the record origin so that we can fire the corresponding app event hook(s) (`OnModelBeforeCreate`, `OnModelBeforeUpdate`, etc.).

One obvious workaround could be to execute a SELECT statement prior the action but that may require wrapping each "upsert" (SELECT + INSERT/UPDATE) in a mixed read+write transaction which could result in poor performance.

For now this request remains out of the scope but it exists in my never ending to do so I may revisit it in the future.

Also note that the above mentioned bulk create/update/delete Web APIs are already implemented as part of the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355).

**Reply by lukepighetti** [reply]
ahh yes I see what you mean. I just meant a high level strategy for allowing upserts, not specific to using SQLite `UPSERT` clause

**Reply by ganigeorgiev** [reply]
_Just for info in case someone stumble here: client-side upsert support was implemented as part of the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) and it will be available with the next v0.23.0 release soon (there are no ETAs yet). If the submitted request body has an id and that id exists internally we will invoke the update handler, otherwise - the create. All related update/create hooks are fired by their corresponding handlers._

_**The upsert functionality works only as part of the new batch/transactional Web APIs and it is limited within the max batch/transaction execution time and its max body size**, both configurable from the UI settings. Or in other words, similar to the batch update and create, the upsert action will not be suitable for uploading large files as this will require keeping the transaction lock for too long and preventing other users interacting with the database._
